services/motion/builtin/move_request.go

```diff
@@ -85,12 +90,89 @@ func (mr *moveRequest) deviatedFromPlan(ctx context.Context, waypoints [][]refer
 	if err != nil {
 		return false, err
 	}
-	mr.planRequest.Logger.Debug("deviation from plan: %v", errorState.Point())
+	mr.planRequest.Logger.Debugf("deviation from plan: %v", errorState.Point())
 	return errorState.Point().Norm() > mr.config.PlanDeviationMM, nil
 }
 
 func (mr *moveRequest) obstaclesIntersectPlan(ctx context.Context, waypoints [][]referenceframe.Input, waypointIndex int) (bool, error) {
-	// TODO(RSDK-4507): implement this function
+	var plan motionplan.Plan
+	// We only care to check against waypoints we have not reached yet.
+	for _, inputs := range waypoints[waypointIndex:] {
+		input := make(map[string][]referenceframe.Input)
+		input[mr.kinematicBase.Name().Name] = inputs
+		plan = append(plan, input)
+	}
+	for _, cvMap := range mr.cameraVisionPairs {
+		for camName, visSrvc := range cvMap {
+			// get detections from vision service
+			detections, err := visSrvc.GetObjectPointClouds(ctx, camName.Name, nil)
+			if err != nil && strings.Contains(err.Error(), "does not implement a 3D segmenter") {
+				mr.planRequest.Logger.Debugf("cannot call GetObjectPointClouds on %q as it does not implement a 3D segmenter", visSrvc.Name())
+				break
+			} else if err != nil {
+				return false, err
+			}
+
+			// Note: detections are initially observed from the camera frame but must be transformed to be in
+			// world frame. We cannot use the inputs of the base to transform the detections since they are relative
+
+			// get the current position of the base which we will use to transform the detection into world coordinates
+			currentPosition, err := mr.kinematicBase.CurrentPosition(ctx)
+			if err != nil {
+				return false, err
+			}
+			// get transform of camera to kinematic base origin
+			kinBaseOrigin := referenceframe.NewPoseInFrame(mr.kinematicBase.Name().ShortName(), spatialmath.NewZeroPose())
+			cameraToBase, err := mr.frameSystemService.TransformPose(ctx, kinBaseOrigin, camName.ShortName(), nil)
+			if err != nil {
+				// here we make the assumption the movement sensor is coincident with the base
+				cameraToBase = kinBaseOrigin
+			}
+
+			// where the camera is in world coordinates
+			transformBy := spatialmath.Compose(currentPosition.Pose(), cameraToBase.Pose())
+
+			// Any obstacles specified by the worldstate of the moveRequest will also re-detected here.
+			// There is no need to append the new detections to the existing worldstate.
+			// We can safely build from scratch without excluding any valuable information.
+			geoms := []spatialmath.Geometry{}
+			for _, detection := range detections {
+				geometry := detection.Geometry.Transform(transformBy)
+				geometry.SetLabel("transient" + detection.Geometry.Label())
```

Comment:
Hardcoding names of things like this always sketches me out. Especially since in this case, `detection.Geometry.Label()` is probably going to frequently be an empty string.

I'd prefer to make this more robust against accidental collisions with something like
```
label := cameraName + "_transientObstacle_" + obstacleNumber
if detection.Geometry.Label() != "" {
    label += "_" + detection.Geometry.Label()
}
geometry.SetLabel(label)
```